### PR TITLE
mgr/snap_schedule: restore yearly spec to lowercase y

### PR DESCRIPTION
**Pull request by mchangir**
Fixes: https://tracker.ceph.com/issues/65350
Signed-off-by: Milind Changire <mchangir@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Review by joscollin — APPROVED** [review]
Looks Good

**Comment by vshankar** [comment]
@mchangir what happens to existing shedules (iff `Y` got used to set a schedule)?

**Comment by vshankar** [comment]
> > @mchangir what happens to existing shedules (iff `Y` got used to set a schedule)?
> 
> haven't explicitly tested this aspect but as long as the mgr isn't restarted, the timer would be set for 1 year and the retention would continue to collect snapshots
> 
> * as soon as the mgr is started with the new bits to restore the yearly spec back to lowercase 'y', then the timer arming would fail since there would be no multiplier found for uppercase 'Y'
> * if uppercase 'Y' was the only retention spec, then the new bits which restores the yearly spec back to lowercase 'y' would result in no valid retention spec and all collected yearly snapshots would be purged
> 
> All of this also depends on the start time from which the year starts. With no explicit start time, the yearly snapshot would fall on Jan 01 00:00 of every year.

So, if the mgr is updated and a spec using `Y` exist in the DB, then that's not a valid spec, right?

**Comment by mchangir** [comment]
> So, if the mgr is updated and a spec using `Y` exist in the DB, then that's not a valid spec, right?

that's right

**Review by rishabh-d-dave — APPROVED** [review]

**Comment by vshankar** [comment]
> > So, if the mgr is updated and a spec using `Y` exist in the DB, then that's not a valid spec, right?
> 
> that's right

Did that change make it to a release? We have to do _something_ about it if the change made it to a release.

**Comment by mchangir** [comment]
> > > So, if the mgr is updated and a spec using `Y` exist in the DB, then that's not a valid spec, right?
> > 
> > 
> > that's right
> 
> Did that change make it to a release? We have to do _something_ about it if the change made it to a release.

yes, the change has made it to:
* reef - 18.2.2
* squid - 19.0.0

**Comment by mchangir** [comment]
The change has been backported to quincy as well, but so far there's been no quincy release with the change.

**Comment by vshankar** [comment]
> > > > So, if the mgr is updated and a spec using `Y` exist in the DB, then that's not a valid spec, right?
> > > 
> > > 
> > > that's right
> > 
> > 
> > Did that change make it to a release? We have to do _something_ about it if the change made it to a release.
> 
> yes, the change has made it to:
> 
> * reef - 18.2.2
> * squid - 19.0.0

We can fix squid as it's still RC. We need a to get this fixed in reef.

**Comment by mchangir** [comment]
> > > > > So, if the mgr is updated and a spec using `Y` exist in the DB, then that's not a valid spec, right?
> > > > 
> > > > 
> > > > that's right
> > > 
> > > 
> > > Did that change make it to a release? We have to do _something_ about it if the change made it to a release.
> > 
> > 
> > yes, the change has made it to:
> > 
> > * reef - 18.2.2
> > * squid - 19.0.0
> 
> We can fix squid as it's still RC. We need a to get this fixed in reef.

IMO this PR along with relevant Release Notes to bring this to users' attention and requesting users to abstain from using 'Y' and instead reverting to use 'y' if it has been used in active retention spec would be sufficient.

File: src/pybind/mgr/snap_schedule/fs/schedule.py
Comment (chrisphoffman) [inline src/pybind/mgr/snap_schedule/fs/schedule.py:1]: Don't forget to change formatting for year here
https://github.com/ceph/ceph/blob/2685bc0aefbe7834e5d37f95e36d1e461ad2f48f/src/pybind/mgr/snap_schedule/fs/schedule.py#L111
and
https://github.com/ceph/ceph/blob/2685bc0aefbe7834e5d37f95e36d1e461ad2f48f/src/pybind/mgr/snap_schedule/fs/schedule.py#L413-L415

Comment (mchangir) [reply src/pybind/mgr/snap_schedule/fs/schedule.py:1]: @chrisphoffman thanks for catching this

Comment (vshankar) [reply src/pybind/mgr/snap_schedule/fs/schedule.py:1]: 👍

**Review by vshankar — APPROVED** [review]

**Review by rishabh-d-dave — COMMENTED** [review]
This PR branch has been merged in the testing branch now.

File: src/pybind/mgr/snap_schedule/fs/schedule.py
Comment (mchangir) [reply src/pybind/mgr/snap_schedule/fs/schedule.py:1]: @neesingh-rh thanks .. this is fixed now

**Review by rishabh-d-dave — COMMENTED** [review]
QA run was successful for this PR. QA run tooks some time due infra issues (specifically CentOS 9 issue) that were persistent in Sepia lab. Besides, there were new failues in the QA run, both related and unrelated to PRs on the testing branch.

However, since this PR has gone through change since it was merged in testing branch, it needs to be tested again.

**Review by rishabh-d-dave — COMMENTED** [review]
QA run was successful - https://tracker.ceph.com/projects/cephfs/wiki/main#3-May-2024.

Testing took more time than expected because there were 25-30 new failures. Most of them caused by a PR in the testing branch but these were resolved on removing that PR.

**Comment by rishabh-d-dave** [comment]
make check result has been deleted - https://jenkins.ceph.com/job/ceph-pull-requests/133444/

**Comment by rishabh-d-dave** [comment]
Waiting for `make check`  to finish so that I can proceed to merge.

**Comment by rishabh-d-dave** [comment]
`make check` ran successfully - https://jenkins.ceph.com/job/ceph-pull-requests/134377/. Proceeding to merge.

**Pull request merged**